### PR TITLE
Add Alignment Imposition

### PR DESCRIPTION
**Pull request by AD2605**
Imposes an alignment requirements on the vectors passed to SYCL' vector class' `load/store` method.
Fixes the issue described in #508

### CODE REVIEW

File: adoc/chapters/programming_interface.adoc [hunk 1]
```diff
@@ -17218,15 +17218,15 @@ a@
 template <access::address_space AddressSpace, access::decorated IsDecorated>
 void load(size_t offset, multi_ptr<const DataT, AddressSpace, IsDecorated> ptr)
 ----
-   a@ Loads the values at the address of [code]#ptr# offset in elements of type [code]#DataT# by [code]#NumElements * offset#, into the components of this SYCL [code]#vec#.
+   a@ Loads the values at the address of [code]#ptr# offset in elements of type [code]#DataT# by [code]#NumElements * offset#, into the components of this SYCL [code]#vec#. [code]#vec# must be aligned to the size of the data type of this SYCL vec.
```
Comment (gmlueck) [inline adoc/chapters/programming_interface.adoc:17221]: Should be something like:

> The `ptr` must be aligned to `alignof(DataT)`.

**Comment by gmlueck** [comment]
@AD2605: I think we are waiting for you to update this PR.

File: adoc/chapters/programming_interface.adoc (hunk 1 above)
Comment (AD2605) [reply adoc/chapters/programming_interface.adoc]: @gmlueck  sorry for the delay
Just pushed the necessary change. Let me know if any other changes would be required

**Review by AerialMantis — APPROVED** [review]

File: adoc/chapters/programming_interface.adoc (hunk 1 above)
Comment (gmlueck) [reply adoc/chapters/programming_interface.adoc]: Thanks, that looks good.

**Review by keryell — COMMENTED** [review]
The addendum sounds good, but I am still trying to understand the original sentence.

**Comment by AD2605** [comment]
@keryell do you mean the original sentence before the review comment this PR was proposing to add ?
Or the motivation behind this PR ?

**Comment by gmlueck** [comment]
@AD2605 will improve first sentence in description also.

**Comment by AD2605** [comment]
@gmlueck  @keryell  updated the original description of the load/store methods. 
Lemme know if it's clear and concise, and if any further change might be required

File: adoc/chapters/programming_interface.adoc [hunk 2]
```diff
@@ -17218,15 +17218,15 @@ a@
 template <access::address_space AddressSpace, access::decorated IsDecorated>
 void load(size_t offset, multi_ptr<const DataT, AddressSpace, IsDecorated> ptr)
 ----
-   a@ Loads the values at the address of [code]#ptr# offset in elements of type [code]#DataT# by [code]#NumElements * offset#, into the components of this SYCL [code]#vec#.
+   a@ Loads [code]#NumElements# elements from consecutive addresses, with the starting address determined by incrementing [code]#ptr# by [code]#offset# provided in terms of number of elements of type [code]#DataT#,  into the components of this SYCL [code]#vec#. The [code]#ptr# must be aligned to [code]#alignof(DataT)#.
```
Comment (gmlueck) [inline adoc/chapters/programming_interface.adoc:17221]: This is not how DPC++ behaves.  Notice how the original wording said:

> offset in elements of type `DataT` by `NumElements * offset`

I think the intent is that `offset` is multiplied by the size of the entire vector (i.e. `NumElements * offset`).  This is how DPC++ behaves.  Therefore, I think the wording should be like this instead:

> Loads `NumElements` elements into the components of `vec`.  These elements are loaded from consecutive addresses, where the starting address is computed by adding `offset * NumElements * sizeof(DataT)` bytes to the address specified by `ptr`.  The `ptr` must be aligned to `alignof(DataT)`.

**Review by gmlueck — APPROVED** [review]
Thanks for your persistence.  This looks good!

**Comment by gmlueck** [comment]
BTW, feel free to add yourself to the list of Contributors in "adoc/acknowledgements.adoc".

**Comment by AD2605** [comment]
Done

File: adoc/chapters/acknowledgements.adoc [hunk 3]
```diff
@@ -90,6 +90,7 @@
   * Peter Thoman, University of Innsbruck
   * Biagio Cosenza, University of Salerno
   * Paul Preney, University of Windsor
+  * Atharva Dubey, Codeplay
```
Comment (gmlueck) [inline adoc/chapters/acknowledgements.adoc:93]: The list is grouped by organization and then sorted by last name within each organization.  Can you move your name to appropriate place in the Codeplay section?

Comment (AD2605) [reply adoc/chapters/acknowledgements.adoc:93]: Thanks for letting me know, will do.

Comment (AD2605) [reply adoc/chapters/acknowledgements.adoc]: Done

**Review by nliber — APPROVED** [review]

**Review by keryell — CHANGES_REQUESTED** [review]
Almost there!

File: adoc/chapters/programming_interface.adoc [hunk 4]
```diff
@@ -17230,15 +17230,15 @@ a@
 template <access::address_space AddressSpace, access::decorated IsDecorated>
 void load(size_t offset, multi_ptr<const DataT, AddressSpace, IsDecorated> ptr)
 ----
-   a@ Loads the values at the address of [code]#ptr# offset in elements of type [code]#DataT# by [code]#NumElements * offset#, into the components of this SYCL [code]#vec#.
+   a@ Loads [code]#NumElements# elements into the components of this SYCL [code]#vec#. These elements are loaded from consecutive addresses, where the starting address is computed by adding [code]#offset * NumElements * sizeof(DataT)# bytes to the address specified by the [code]#ptr#. The ptr must be aligned to [code]#alignof(DataT)#.
```
[the remark this answers is not in the document]
Comment (AD2605) [reply adoc/chapters/programming_interface.adoc]: Thanks, added `ptr` inside the code section

**Review by keryell — APPROVED** [review]
Thanks!

**Comment by tomdeakin** [comment]
WG approved, pleased merge

**Pull request merged**